pkg/util/strategicpatch/patch.go

```diff
@@ -1447,3 +1453,34 @@ func GetServerSupportedSMPatchVersion(discoveryClient discovery.DiscoveryInterfa
 	}
 	return Unknown, fmt.Errorf("The version is too old: %v\n", serverVersion)
 }
+
+// GuidedJsonMerge merges two JSON formatted items by following the structure of dataStruct's type. We use
+// dataStruct's type to respect strategicMerge guidelines and ensure the structure of the JSON items
+// (e.g. no fields are missing, no additional fields exist, etc)
+func GuidedJsonMerge(j1, j2 []byte, dataStruct interface{}) ([]byte, error) {
+	j1Map := map[string]interface{}{}
+	if len(j1) > 0 {
+		if err := json.Unmarshal(j1, &j1Map); err != nil {
+			return nil, fmt.Errorf("Invalid JSON document")
+		}
+	}
+
+	j2Map := map[string]interface{}{}
+	if len(j2) > 0 {
+		if err := json.Unmarshal(j2, &j2Map); err != nil {
+			return nil, fmt.Errorf("Invalid JSON document")
+		}
+	}
+
+	t, err := getTagStructType(dataStruct)
+	if err != nil {
+		return nil, err
+	}
+
+	mergedMap, err := mergeMap(j1Map, j2Map, t, false)
```

Comment:
Good question on #3.  By loading are you referring to the defaulting?  I can't remember if we are still doing defaulting of fields in kubectl.  @AdoHe Do you know if we ever do the server-side defaulting in kubectl, or do we just do the round-tripping of objects through structs?

@andronat I would put your list of 3 differences as a comment above the mergemap line since it isn't obvious.  I'd also rename this function to be "MergeExplicitNulls".  I know this makes it less generic, but makes it more obvious what the intentions are and we aren't using it anywhere else.